src/util/GenericEventEmitter.ts

```diff
@@ -0,0 +1,36 @@
+import { EventEmitter } from 'events';
+
+/**
+ * A typed interface of {@link EventEmitter}.
+ *
+ * Use the `&` operator to combine multiple event/function pairs into a single event emitter.
+ * The result needs to be a type and not an interface because of https://github.com/microsoft/TypeScript/issues/16936.
+ *
+ * Use the {@link createGenericEventEmitterClass} function to generate an event emitter class with the correct typings
+ * in case {@link EventEmitter} needs to be extended.
+ */
+export interface GenericEventEmitter<TEvent extends string | symbol, TFunc extends (...args: any[]) => void>
+  extends EventEmitter {
+  addListener: (event: TEvent, listener: TFunc) => this;
+  on: (event: TEvent, listener: TFunc) => this;
+  once: (event: TEvent, listener: TFunc) => this;
+  removeListener: (event: TEvent, listener: TFunc) => this;
+  off: (event: TEvent, listener: TFunc) => this;
+  removeAllListeners: (event: TEvent) => this;
+  listeners: (event: TEvent) => TFunc[];
+  rawListeners: (event: TEvent) => TFunc[];
+  emit: (event: TEvent, ...args: Parameters<TFunc>) => boolean;
+  listenerCount: (event: TEvent) => number;
+  prependListener: (event: TEvent, listener: TFunc) => this;
+  prependOnceListener: (event: TEvent, listener: TFunc) => this;
+  eventNames: () => TEvent[];
+}
+
+/**
+ * Creates a class that is an implementation of {@link EventEmitter}
+ * but with specific typings based on {@link GenericEventEmitter}.
+ * Useful in case a class needs to extend {@link EventEmitter} and wants specific internal typings.
+ */
+export function createGenericEventEmitterClass<T extends EventEmitter>(): (new() => T) {
```

Comment:
`GenericEventEmitter<T>` is the typed version of `EventEmitter`. The interface that is. The problem is classes that extend `EventEmitter`, such as `MonitoringStore`. if they still extend `EventEmitter` you would not have the correct typings internally. So `this.emit(...)` would not check for correct typings. Even if the class is defined as `class MonitoringStore extends EventEmitter implements GenericEventEmitter<...>`. The only way to have the internal typings be correct is to extend a class that has correct typings (or rewrite all the functions internally).

So what we need is a class that is an implementation of EventEmitter, but actually says its interface is `GenericEventEmitter<T>`. The above function generates such a class.

It might be possible to create a class with generics that can be extended immediately. All EventEmitter functions would have to be reimplemented though to cast them as TS would complain the output is not correct (although we know it is, which is why we can create this interface). But then that wouldn't allow us to combine multiple `GenericEventEmitter` interfaces like we do for the activity emitter.

One of the problems is that there are less casting opportunities when extending and there is a clear difference between classes/implementations and interfaces/types/descriptions.

... or there is a different solution I didn't think of yet 😄 .

Reply:
`export class EventEmitter<T>` where the constructor returns `new events.EventEmitter`? But yeah, you'd still need to type out the empty methods I presume.
And I see the problem of combining things indeed.